sunscreen_math/src/poly/mod.rs

```diff
@@ -1,32 +1,179 @@
+    ///
+    /// # Remarks
+    /// Variable time
+    fn eq(&self, other: &Self) -> bool {
+        // Need to handle zero polynomial specially, as calling degree will panic.
+        let lhs_is_zero = self.vartime_is_zero();
+        let rhs_is_zero = other.vartime_is_zero();
+
+        if lhs_is_zero || rhs_is_zero {
+            return lhs_is_zero && rhs_is_zero;
+        }
+
+        let lhs_degree = self.vartime_degree();
+        let rhs_degree = other.vartime_degree();
+
+        if lhs_degree != rhs_degree {
+            return false;
+        }
+
+        for i in 0..lhs_degree {
+            if self.coeffs[i] != other.coeffs[i] {
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+impl<R> Polynomial<R>
+where
+    R: Ring,
 {
     /// Construct a new polynomial with the given coefficients.
-    pub fn new(coeffs: &[T]) -> Self {
+    pub fn new(coeffs: &[R]) -> Self {
         Self {
             coeffs: coeffs.to_owned(),
         }
     }
+
+    /// Evaluate the polynomial at x.
+    pub fn evaluate(&self, x: &R) -> R {
+        let mut eval = R::zero();
+        let mut cur_pow = R::one();
+
+        for i in &self.coeffs {
+            eval = eval + i.clone() * &cur_pow;
+            cur_pow = cur_pow.clone() * x;
+        }
+
+        eval
+    }
+
+    /// Returns the degree of the polynomial
+    ///
+    /// # Remarks
+    /// Runtime varies depending on the number of leading zeros.
+    ///
+    /// # Panics
+    /// The degree of the zero polynomial is undefined, and thus this function will
+    /// panic.
+    pub fn vartime_degree(&self) -> usize {
+        for (i, coeff) in self.coeffs.iter().rev().enumerate() {
+            if !coeff.vartime_is_zero() {
+                return self.coeffs.len() - i - 1;
+            }
+        }
+
+        panic!("Zero polynomial has undefined degree.");
+    }
+
+    /// Computes the quotient and remainder of `self / rhs`.
+    ///
+    /// # Remarks
+    /// Runtime is variable except for very restricted use cases.
+    /// This function will be constant time so long as:
+    /// * Neither the numerator nor denominator have leading zeros.
+    /// * The numerator is always of higher degree than the denominator
+    /// * The numerator's and denominator's degrees are fixed across invocations
+    /// * The inner type R supports constant time subtraction and multiplication.
+    ///
+    /// In order for polynomial division to work in a ring, `rhs` has restrictions.
+    /// Specifically, the highest order non-zero coefficient must be 1 so as to avoid
+    /// inverse operations. While multiplicative inverses are not guaranteed to exist
+    /// for a [`Ring`] element, the inverse of one will always be one.
+    ///
+    /// # Panics
+    /// If the divisor's leading non-zero coefficient isn't one.
+    ///
+    /// If rhs is the zero polynomial.
+    pub fn vartime_div_rem_restricted_rhs(&self, rhs: &Self) -> (Self, Self) {
```

Comment:
maybe want to add a test for this function - unless it's covered elsewhere and I missed it

Reply:
nvm - this is tested below

Reply:
Check the end of the file.